forage-android/src/main/java/com/joinforage/forage/android/collect/AbstractVaultSubmitter.kt

```diff
@@ -0,0 +1,182 @@
+
+        // ========= USED FOR REPORTING IMPORTANT METRICS =========
+        val proxyResponseMonitor = VaultProxyResponseMonitor(
+            vault = vaultType,
+            userAction = userAction,
+            metricsLogger = logger
+        )
+        proxyResponseMonitor
+            .setPath(path)
+            .setMethod(HTTPMethod.POST.toString())
+            .start()
+
+        val vaultProxyRequest = buildProxyRequest(
+            encryptionKey = encryptionKey,
+            idempotencyKey = idempotencyKey,
+            merchantId = merchantId,
+            vaultToken = vaultToken
+        ).setPath(path)
+        val forageResponse = submitProxyRequest(vaultProxyRequest = vaultProxyRequest)
+        proxyResponseMonitor.end()
+
+        // FNS requirement to clear the PIN after each submission
+        foragePinEditText.clearText()
+
+        val httpStatusCode = when (forageResponse) {
+            is ForageApiResponse.Success -> 200
+            is ForageApiResponse.Failure -> forageResponse.errors[0].httpStatusCode
+        }
+        proxyResponseMonitor.setHttpStatusCode(httpStatusCode).logResult()
+        // ==========================================================
+
+        return forageResponse
+    }
+
+    override fun getVaultType(): VaultType {
+        return vaultType
+    }
+
+    // abstract methods
+    internal abstract fun parseEncryptionKey(encryptionKeys: EncryptionKeys): String
+    internal abstract suspend fun submitProxyRequest(vaultProxyRequest: VaultProxyRequest): ForageApiResponse<String>
+    internal abstract fun getVaultToken(paymentMethod: PaymentMethod): String?
+
+    internal abstract fun toVaultErrorOrNull(vaultResponse: VaultResponse): ForageApiResponse.Failure?
+    abstract fun toForageErrorOrNull(vaultResponse: VaultResponse): ForageApiResponse.Failure?
+    abstract fun toForageSuccessOrNull(vaultResponse: VaultResponse): ForageApiResponse.Success<String>?
+    abstract fun parseVaultError(vaultResponse: VaultResponse): String
+
+    // concrete methods
+    protected open fun buildProxyRequest(
+        encryptionKey: String,
+        idempotencyKey: String,
+        merchantId: String,
+        vaultToken: String
+    ) = VaultProxyRequest.emptyRequest()
+        .setHeader(ForageConstants.Headers.X_KEY, encryptionKey)
+        .setHeader(ForageConstants.Headers.MERCHANT_ACCOUNT, merchantId)
+        .setHeader(ForageConstants.Headers.IDEMPOTENCY_KEY, idempotencyKey)
+        .setHeader(ForageConstants.Headers.TRACE_ID, logger.getTraceIdValue())
+        .setToken(vaultToken)
+
+    protected fun pickVaultTokenByIndex(paymentMethod: PaymentMethod, index: Int): String? {
+        val tokensString = paymentMethod.card.token
+        val tokensList = tokensString.split(TOKEN_DELIMITER)
+
+        // VGS is always the first token in the string
+        val noTokenStoredInVault = tokensList.size <= index
+        if (noTokenStoredInVault) return null
+
+        // grab BasisTheory token
```

Comment:
Adjust comment this is a generic function.

Reply:
28e4362a995f7e3c5f76225562da1dc2faead80c